AppleSource/shell_cmds-198/find/find.1
.\" Copyright (c) 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)find.1	8.7 (Berkeley) 5/9/95
.\" $FreeBSD: src/usr.bin/find/find.1,v 1.91 2011/09/28 18:53:36 ed Exp $
.\"
.Dd September 28, 2011
.Dt FIND 1
.Os
.Sh NAME
.Nm find
.Nd walk a file hierarchy
.Sh SYNOPSIS
.Nm
.Op Fl H | Fl L | Fl P
.Op Fl EXdsx
.Op Fl f Ar path
.Ar path ...
.Op Ar expression
.Nm
.Op Fl H | Fl L | Fl P
.Op Fl EXdsx
.Fl f Ar path
.Op Ar path ...
.Op Ar expression
.Sh DESCRIPTION
The
.Nm
utility recursively descends the directory tree for each
.Ar path
listed, evaluating an
.Ar expression
(composed of the
.Dq primaries
and
.Dq operands
listed below) in terms
of each file in the tree.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl E
Interpret regular expressions followed by
.Ic -regex
and
.Ic -iregex
primaries as extended (modern) regular expressions rather than basic
regular expressions (BRE's).
The
.Xr re_format 7
manual page fully describes both formats.
.It Fl H
Cause the file information and file type (see
.Xr stat 2 )
returned for each symbolic link specified on the command line to be
those of the file referenced by the link, not the link itself.
If the referenced file does not exist, the file information and type will
be for the link itself.
File information of all symbolic links not on
the command line is that of the link itself.
.It Fl L
Cause the file information and file type (see
.Xr stat 2 )
returned for each symbolic link to be those of the file referenced by the
link, not the link itself.
If the referenced file does not exist, the file information and type will
be for the link itself.
.Pp
This option is equivalent to the deprecated
.Ic -follow
primary.
.It Fl P
Cause the file information and file type (see
.Xr stat 2 )
returned for each symbolic link to be those of the link itself.
This is the default.
.It Fl X
Permit
.Nm
to be safely used in conjunction with
.Xr xargs 1 .
If a file name contains any of the delimiting characters used by
.Xr xargs 1 ,
a diagnostic message is displayed on standard error, and the file
is skipped.
The delimiting characters include single
.Pq Dq Li " ' "
and double
.Pq Dq Li " \*q "
quotes, backslash
.Pq Dq Li \e ,
space, tab and newline characters.
.Pp
However, you may wish to consider the
.Fl print0
primary in conjunction with
.Dq Nm xargs Fl 0
as an effective alternative.
.It Fl d
Cause
.Nm
to perform a depth-first traversal, i.e., directories
are visited in post-order and all entries in a directory will be acted
on before the directory itself.
By default,
.Nm
visits directories in pre-order, i.e., before their contents.
Note, the default is
.Em not
a breadth-first traversal.
.Pp
This option is equivalent to the
.Ic -depth
primary of
.St -p1003.1-2001 .
The
.Fl d
option
can be useful when
.Nm
is used with
.Xr cpio 1
to process files that are contained in directories with unusual permissions.
It ensures that you have write permission while you are placing files in a
directory, then sets the directory's permissions as the last thing.
.It Fl f
Specify a file hierarchy for
.Nm
to traverse.
File hierarchies may also be specified as the operands immediately
following the options.
.It Fl s
Cause
.Nm
to traverse the file hierarchies in lexicographical order,
i.e., alphabetical order within each directory.
Note:
.Ql find -s
and
.Ql "find | sort"
may give different results.
.It Fl x
Prevent
.Nm
from descending into directories that have a device number different
than that of the file from which the descent began.
.Pp
This option is equivalent to the deprecated
.Ic -xdev
primary.
.El
.Sh PRIMARIES
.Pp
All primaries which take a numeric argument allow the number to be
preceded by a plus sign
.Pq Dq Li +
or a minus sign
.Pq Dq Li - .
A preceding plus sign means
.Dq more than n ,
a preceding minus sign means
.Dq less than n
and neither means
.Dq exactly n .
.Bl -tag -width indent
.It Ic -Bmin Ar n
True if the difference between the time of a file's inode creation
and the time
.Nm
was started, rounded up to the next full minute, is
.Ar n
minutes.
.It Ic -Bnewer Ar file
Same as
.Ic -newerBm .
.It Ic -Btime Ar n Ns Op Cm smhdw
If no units are specified, this primary evaluates to
true if the difference between the time of a file's inode creation
and the time
.Nm
was started, rounded up to the next full 24-hour period, is
.Ar n
24-hour periods.
.Pp
If units are specified, this primary evaluates to
true if the difference between the time of a file's inode creation
and the time
.Nm
was started is exactly
.Ar n
units.
Please refer to the
.Ic -atime
primary description for information on supported time units.
.It Ic -acl
May be used in conjunction with other primaries to locate
files with extended ACLs.
See
.Xr acl 3
for more information.
.It Ic -amin Ar n
True if the difference between the file last access time and the time
.Nm
was started, rounded up to the next full minute, is
.Ar n
minutes.
.It Ic -anewer Ar file
Same as
.Ic -neweram .
.It Ic -atime Ar n Ns Op Cm smhdw
If no units are specified, this primary evaluates to
true if the difference between the file last access time and the time
.Nm
was started, rounded up to the next full 24-hour period, is
.Ar n
24-hour periods.
.Pp
If units are specified, this primary evaluates to
true if the difference between the file last access time and the time
.Nm
was started is exactly
.Ar n
units.
Possible time units are as follows:
.Pp
.Bl -tag -width indent -compact
.It Cm s
second
.It Cm m
minute (60 seconds)
.It Cm h
hour (60 minutes)
.It Cm d
day (24 hours)
.It Cm w
week (7 days)
.El
.Pp
Any number of units may be combined in one
.Ic -atime
argument, for example,
.Dq Li "-atime -1h30m" .
Units are probably only useful when used in conjunction with the
.Cm +
or
.Cm -
modifier.
.It Ic -cmin Ar n
True if the difference between the time of last change of file status
information and the time
.Nm
was started, rounded up to the next full minute, is
.Ar n
minutes.
.It Ic -cnewer Ar file
Same as
.Ic -newercm .
.It Ic -ctime Ar n Ns Op Cm smhdw
If no units are specified, this primary evaluates to
true if the difference between the time of last change of file status
information and the time
.Nm
was started, rounded up to the next full 24-hour period, is
.Ar n
24-hour periods.
.Pp
If units are specified, this primary evaluates to
true if the difference between the time of last change of file status
information and the time
.Nm
was started is exactly
.Ar n
units.
Please refer to the
.Ic -atime
primary description for information on supported time units.
.It Ic -d
Same as 
.Ic depth .
GNU find implements this as a primary in mistaken emulation of
.Fx
.Xr find 1 .
.It Ic -delete
Delete found files and/or directories.
Always returns true.
This executes
from the current working directory as
.Nm
recurses down the tree.
It will not attempt to delete a filename with a
.Dq Pa /
character in its pathname relative to
.Dq Pa \&.
for security reasons.
Depth-first traversal processing is implied by this option.
Following symlinks is incompatible with this option.
.It Ic -depth
Always true;
same as the
.Fl d
option.
.It Ic -depth Ar n
True if the depth of the file relative to the starting point of the traversal
is
.Ar n .
.It Ic -empty
True if the current file or directory is empty.
.It Ic -exec Ar utility Oo Ar argument ... Oc Li \&;
True if the program named
.Ar utility
returns a zero value as its exit status.
Optional
.Ar arguments
may be passed to the utility.
The expression must be terminated by a semicolon
.Pq Dq Li \&; .
If you invoke
.Nm
from a shell you may need to quote the semicolon if the shell would
otherwise treat it as a control operator.
If the string
.Dq Li {}
appears anywhere in the utility name or the
arguments it is replaced by the pathname of the current file.
.Ar Utility
will be executed from the directory from which
.Nm
was executed.
.Ar Utility
and
.Ar arguments
are not subject to the further expansion of shell patterns
and constructs.
.It Ic -exec Ar utility Oo Ar argument ... Oc Li {} +
Same as
.Ic -exec ,
except that
.Dq Li {}
is replaced with as many pathnames as possible for each invocation of
.Ar utility .
This behaviour is similar to that of
.Xr xargs 1 .
.It Ic -execdir Ar utility Oo Ar argument ... Oc Li \&;
The
.Ic -execdir
primary is identical to the
.Ic -exec
primary with the exception that
.Ar utility
will be executed from the directory that holds
the current file.
The filename substituted for
the string
.Dq Li {}
is not qualified.
.It Ic -execdir Ar utility Oo Ar argument ... Oc Li {} +
Same as
.Ic -execdir ,
except that
.Dq Li {}
is replaced with as many pathnames as possible for each invocation of
.Ar utility .
This behaviour is similar to that of
.Xr xargs 1 .
.It Ic -flags Oo Cm - Ns | Ns Cm + Oc Ns Ar flags , Ns Ar notflags
The flags are specified using symbolic names (see
.Xr chflags 1 ) .
Those with the
.Qq Li no
prefix (except
.Qq Li nodump )
are said to be
.Ar notflags .
Flags in
.Ar flags
are checked to be set, and flags in
.Ar notflags
are checked to be not set.
Note that this is different from
.Ic -perm ,
which only allows the user to specify mode bits that are set.
.Pp
If flags are preceded by a dash
.Pq Dq Li - ,
this primary evaluates to true
if at least all of the bits in
.Ar flags
and none of the bits in
.Ar notflags
are set in the file's flags bits.
If flags are preceded by a plus
.Pq Dq Li + ,
this primary evaluates to true
if any of the bits in
.Ar flags
is set in the file's flags bits,
or any of the bits in
.Ar notflags
is not set in the file's flags bits.
Otherwise,
this primary evaluates to true
if the bits in
.Ar flags
exactly match the file's flags bits,
and none of the
.Ar flags
bits match those of
.Ar notflags .
.It Ic -fstype Ar type
True if the file is contained in a file system of type
.Ar type .
The
.Xr lsvfs 1
command can be used to find out the types of file systems
that are available on the system.
In addition, there are two pseudo-types,
.Dq Li local
and
.Dq Li rdonly .
The former matches any file system physically mounted on the system where
the
.Nm
is being executed and the latter matches any file system which is
mounted read-only.
.It Ic -gid Ar gname
The same thing as
.Ar -group Ar gname 
for compatibility with GNU find.
GNU find imposes a restriction that
.Ar gname 
is numeric, while
.Xr find 1 
does not.
.It Ic -group Ar gname
True if the file belongs to the group
.Ar gname .
If
.Ar gname
is numeric and there is no such group name, then
.Ar gname
is treated as a group ID.
.It Ic -ignore_readdir_race
This option is for GNU find compatibility and is ignored.
.It Ic -ilname Ar pattern
Like
.Ic -lname ,
but the match is case insensitive.
This is a GNU find extension.
.It Ic -iname Ar pattern
Like
.Ic -name ,
but the match is case insensitive.
.It Ic -inum Ar n
True if the file has inode number
.Ar n .
.It Ic -ipath Ar pattern
Like
.Ic -path ,
but the match is case insensitive.
.It Ic -iregex Ar pattern
Like
.Ic -regex ,
but the match is case insensitive.
.It Ic -iwholename Ar pattern
The same thing as 
.Ic -ipath ,
for GNU find compatibility.
.It Ic -links Ar n
True if the file has
.Ar n
links.
.It Ic -lname Ar pattern
Like
.Ic -name ,
but the contents of the symbolic link are matched instead of the file
name.
Note that this only matches broken symbolic links
if symbolic links are being followed.
This is a GNU find extension.
.It Ic -ls
This primary always evaluates to true.
The following information for the current file is written to standard output:
its inode number, size in 512-byte blocks, file permissions, number of hard
links, owner, group, size in bytes, last modification time, and pathname.
If the file is a block or character special file, the device number
will be displayed instead of the size in bytes.
If the file is a symbolic link, the pathname of the linked-to file will be
displayed preceded by
.Dq Li -> .
The format is identical to that produced by
.Bk -words
.Dq Nm ls Fl dgils .
.Ek
.It Ic -maxdepth Ar n
Always true; descend at most
.Ar n
directory levels below the command line arguments.
If any
.Ic -maxdepth
primary is specified, it applies to the entire expression even if it would
not normally be evaluated.
.Dq Ic -maxdepth Li 0
limits the whole search to the command line arguments.
.It Ic -mindepth Ar n
Always true; do not apply any tests or actions at levels less than
.Ar n .
If any
.Ic -mindepth
primary is specified, it applies to the entire expression even if it would
not normally be evaluated.
.Dq Ic -mindepth Li 1
processes all but the command line arguments.
.It Ic -mmin Ar n
True if the difference between the file last modification time and the time
.Nm
was started, rounded up to the next full minute, is
.Ar n
minutes.
.It Ic -mnewer Ar file
Same as
.Ic -newer .
.It Ic -mount
The same thing as 
.Ic -xdev ,
for GNU find compatibility.
.It Ic -mtime Ar n Ns Op Cm smhdw
If no units are specified, this primary evaluates to
true if the difference between the file last modification time and the time
.Nm
was started, rounded up to the next full 24-hour period, is
.Ar n
24-hour periods.
.Pp
If units are specified, this primary evaluates to
true if the difference between the file last modification time and the time
.Nm
was started is exactly
.Ar n
units.
Please refer to the
.Ic -atime
primary description for information on supported time units.
.It Ic -name Ar pattern
True if the last component of the pathname being examined matches
.Ar pattern .
Special shell pattern matching characters
.Dq ( Li \&[ ,
.Dq Li \&] ,
.Dq Li * ,
and
.Dq Li \&? )
may be used as part of
.Ar pattern .
These characters may be matched explicitly by escaping them with a
backslash
.Pq Dq Li \e .
.It Ic -newer Ar file
True if the current file has a more recent last modification time than
.Ar file .
.It Ic -newer Ns Ar X Ns Ar Y Ar file
True if the current file has a more recent last access time
.Pq Ar X Ns = Ns Cm a ,
inode creation time
.Pq Ar X Ns = Ns Cm B ,
change time
.Pq Ar X Ns = Ns Cm c ,
or modification time
.Pq Ar X Ns = Ns Cm m
than the last access time
.Pq Ar Y Ns = Ns Cm a ,
inode creation time
.Pq Ar Y Ns = Ns Cm B ,
change time
.Pq Ar Y Ns = Ns Cm c ,
or modification time
.Pq Ar Y Ns = Ns Cm m
of
.Ar file .
In addition, if
.Ar Y Ns = Ns Cm t ,
then
.Ar file
is instead interpreted as a direct date specification of the form
understood by
.Xr cvs 1 .
Note that
.Ic -newermm
is equivalent to
.Ic -newer .
.It Ic -nogroup
True if the file belongs to an unknown group.
.It Ic -noignore_readdir_race
This option is for GNU find compatibility and is ignored.
.It Ic -noleaf
This option is for GNU find compatibility.
In GNU find it disables an optimization not relevant to 
.Xr find 1 ,
so it is ignored.
.It Ic -nouser
True if the file belongs to an unknown user.
.It Ic -ok Ar utility Oo Ar argument ... Oc Li \&;
The
.Ic -ok
primary is identical to the
.Ic -exec
primary with the exception that
.Nm
requests user affirmation for the execution of the
.Ar utility
by printing
a message to the terminal and reading a response.
If the response is not affirmative
.Ql ( y
in the
.Dq Li POSIX
locale),
the command is not executed and the
value of the
.Ic -ok
expression is false.
.It Ic -okdir Ar utility Oo Ar argument ... Oc Li \&;
The
.Ic -okdir
primary is identical to the
.Ic -execdir
primary with the same exception as described for the
.Ic -ok
primary.
.It Ic -path Ar pattern
True if the pathname being examined matches
.Ar pattern .
Special shell pattern matching characters
.Dq ( Li \&[ ,
.Dq Li \&] ,
.Dq Li * ,
and
.Dq Li \&? )
may be used as part of
.Ar pattern .
These characters may be matched explicitly by escaping them with a
backslash
.Pq Dq Li \e .
Slashes
.Pq Dq Li /
are treated as normal characters and do not have to be
matched explicitly.
.It Ic -perm Oo Cm - Ns | Ns Cm + Oc Ns Ar mode
The
.Ar mode
may be either symbolic (see
.Xr chmod 1 )
or an octal number.
If the
.Ar mode
is symbolic, a starting value of zero is assumed and the
.Ar mode
sets or clears permissions without regard to the process' file mode
creation mask.
If the
.Ar mode
is octal, only bits 07777
.Pq Dv S_ISUID | S_ISGID | S_ISTXT | S_IRWXU | S_IRWXG | S_IRWXO

CHANGELOG.md
# Version 5.028.1
	- Fixed Cmd+v shortcut
	- HostKey fingerprint as base64 encoding
	- Smoothing the HUD on resize or SplitView
	- Empty default user configuration bugs.
	- Zoom shortcuts changed.
	- Solarized theme fixes.
	
# Version 4.024.2
	- AirTerminals! Put a terminal on your remote AirPlay screen :)
	- Blink shortcuts. Use your external keyboard to move, create, or remove terminals. Configure the trigger too! View settings > Keyboard > Shortcuts for more info.
	- Want to use SmartKeys when an external keyboard is connected? Now you can switch them on and off from configuration.
	- View geometry of the screen from the overlay with each resize.
	- Switch cursor blinking on/off from Appearance settings.

	- Fixed TouchID issues when returning to the app.
	- Fixed cursor blink sequence so that it always starts as ON.
	- Fixed Host Port not resetting properly.
	- Fixed overlay getting covered in landscape mode.

	- Smoother experience thanks to cleanups and improvements.

# Version 3.021.2
	- iCloud Hosts sync. Synchronize hosts between devices. If a Host already has been synced, it provides conflict resolution. No critical data like passwords is saved.
	- Auto Lock. If enabled, when you lock/unlock your device, Blink will also be locked. Passcode and TouchID will be required to unlock the app.
	- Added ARMv7 support. Support for 32 bit devices like iPad 2, 3, iPhone 5, etc... We will publish depending on how well it performs!
	- Added IPv6 support for hosts.
	- Share Public Encryption Keys. You can now share the public key from the Keys section to other apps, like Mail.

	- Updated Fira Code font to v1.204.
	- Improved error checking on Themes and Font uploads. Auto correct if the GH URL is not a raw one.

	- Fixed bug with password not getting saved on host creation.
	- Fixed crash when hitting arrows with landscape keyboard on Plus devices.
	- Rolled back LC_CTYPE enforcement on server.

# Version 2.109 / 3.0
	- NEW Autorepeat for normal keys on external keyboard. Vim users rejoice!
	- NEW Default User for connections based on device.
	- NEW Comments for Public Keys.

	- Security updates to libraries.

	- Fixed critical bugs affecting ssh connections.

# Version 1.031
	- NEW On-screen keyboard with more space for modifiers, FKeys and Cursor keys. Redesigned for more space on the modifiers, and with a central scrollable area that handles more keys. Activate the Alternate keys by taping on the Alt key. And now tap on a modifier to activate it as a normal button, or make a long press to chain different combinations.
	- NEW Add your own Fonts & Themes! More info on (https://github.com/blinksh/fonts) and (https://github.com/blinksh/themes)
	- NEW Multistep authentication. Servers with google authenticator or similar will now connect without problems :)
	- NEW Fira Code font with ligatures included.

	- Added -l parameter to ssh for Hosts.
	- Improved message on how use hosts after adding them.
	- Fixed on-screen arrows.
	- Fixed F0 as F10 on external keyboard.
	- Fixed keys selection problem on settings
	- Fixed ssh-copy-id issue when accessing the host for the first time

# Version 1.019
	- Simplified build process.
	- New README and BUILD instructions

# Version 0.1020
	- "Get In Touch" and "About" sections completed.
	- iOS 10 support.
	- Added exception for iOS distribution.

	- Improved gesture support, specially on iPhone and iOS10.
	- MBProgressHUD updated with patch.
	- Test Terminal fixes for fonts and updates.
	- Disable font ligatures.
	- SSH Session termination fixes.

# Version 0.1010
	- Terminal customization! Customize your terminal with different default themes and fonts. Preview your changes within the Settings Preview section. Save your changes and restore them on each execution. Enjoy!

	- Create an id_rsa on first boot as default key.

	- Swipe and pinch conflicts
	- Copy changes to settings

# Version 0.931
	- Map Caps or Shift taps to Esc key.
	- Cursor keys (Home, End, PgUp, PgDown) with default to Cmd+arrow.
	- Function keys (F1-F10) with default to Cmd+number.
	- Change mapping of any function key to a different combination.

	- Improved placeholders for Hosts configurations.

	- Removed Alt as Esc from initial defaults.

# Version 0.927
	This version should complete the experience in relation to ssh and terminal configurations, mimicking a big part of what you can do in a normal shell:
	- Hosts Configuration. Preconfigure a host parameters, like user, port, key and commands.
	- Connect to a Host by specifying its name. Do "mosh plankton"
	- Overwrite parameters from host configuration from the shell. So carlos@plankton will override the user field on plankton.
	- Default Modifier keys settings changed: Ctrl is Ctrl, and Alt sends Esc. Everything else is undefined.
	- Modifier keys configuration. CAPS as Ctrl or ESC, no problem! Cmd as Ctrl? You have it! Configure everything to you liking.
	- Added secure passwords stored on Keychain to Host Configuration.
	- Exiting the session within the MCP closes the Space.

	- Fixed hang after "exec request accepted". Establishing connections should be smooth now.
	- Fixed adjustments on viewport after rotating the display.
	- Fixed Ctrl + Space sequences.
	- Fixed wrong/unexisting Ctrl sequences.
	- Added Esc+any character support.

	- New settings added to project. Will start to fill up in raw branch.

# Version 0.916
	This version contains many important bug fixes that should improve the experience a lot and reduce crashes to, hopefully, close to zero:
	- Unicode support to Blink shell. Thanks Yury!
	- Support for function keys + modifiers. Now you can do Ctrl/Alt/Shift + Arrows.
	- Improved experience with drag down to close. Let me know what you think :)
	- SSH with improved function keys in ncurses apps. htop, less, etc... should now work perfectly.
	- New help message with version and gestures.
	- Added asterisk, underscore and escape to SmartKeys.
	- Added HockeySDKResources.bundle to project for improved updates.


AppleSource/shell_cmds-198/find/find.c
/*-
 * Copyright (c) 1991, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Cimarron D. Taylor of the University of California, Berkeley.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
#if 0
static char sccsid[] = "@(#)find.c	8.5 (Berkeley) 8/5/94";
#else
#endif
#endif /* not lint */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: src/usr.bin/find/find.c,v 1.23 2010/12/11 08:32:16 joel Exp $");

#include <sys/types.h>
#include <sys/stat.h>

#include <err.h>
#include <errno.h>
#include <fts.h>
#include <regex.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef __APPLE__
//#include <get_compat.h>
#include <unistd.h>
//#else
#define COMPAT_MODE(func, mode) 1
#endif

#include "find.h"
#include "error.h"

#ifdef __APPLE__
static int find_compare(const FTSENT **s1, const FTSENT **s2);
#else /* !__APPLE__ */
static int find_compare(const FTSENT * const *s1, const FTSENT * const *s2);
#endif /* __APPLE__ */

/*
 * find_compare --
 *	tell fts_open() how to order the traversal of the hierarchy. 
 *	This variant gives lexicographical order, i.e., alphabetical
 *	order within each directory.
 */
static int
#ifdef __APPLE__
find_compare(const FTSENT **s1, const FTSENT **s2)
#else /* !__APPLE__ */
find_compare(const FTSENT * const *s1, const FTSENT * const *s2)
#endif /* __APPLE__ */
{

	return (strcoll((*s1)->fts_name, (*s2)->fts_name));
}

/*
 * find_formplan --
 *	process the command line and create a "plan" corresponding to the
 *	command arguments.
 */
PLAN *
find_formplan(char *argv[])
{
	PLAN *plan, *tail, *new;

	/*
	 * for each argument in the command line, determine what kind of node
	 * it is, create the appropriate node type and add the new plan node
	 * to the end of the existing plan.  The resulting plan is a linked
	 * list of plan nodes.  For example, the string:
	 *
	 *	% find . -name foo -newer bar -print
	 *
	 * results in the plan:
	 *
	 *	[-name foo]--> [-newer bar]--> [-print]
	 *
	 * in this diagram, `[-name foo]' represents the plan node generated
	 * by c_name() with an argument of foo and `-->' represents the
	 * plan->next pointer.
	 */
	for (plan = tail = NULL; *argv;) {
		if (!(new = find_create(&argv)))
			continue;
		if (plan == NULL)
			tail = plan = new;
		else {
			tail->next = new;
			tail = new;
		}
	}

	/*
	 * if the user didn't specify one of -print, -ok or -exec, then -print
	 * is assumed so we bracket the current expression with parens, if
	 * necessary, and add a -print node on the end.
	 */
	if (!isoutput) {
		OPTION *p;
		char **argv1 = 0;

		if (plan == NULL) {
			p = lookup_option("-print");
			new = (p->create)(p, &argv1);
			tail = plan = new;
		} else {
			p = lookup_option("(");
			new = (p->create)(p, &argv1);
			new->next = plan;
			plan = new;
			p = lookup_option(")");
			new = (p->create)(p, &argv1);
			tail->next = new;
			tail = new;
			p = lookup_option("-print");
			new = (p->create)(p, &argv1);
			tail->next = new;
			tail = new;
		}
	}

	/*
	 * the command line has been completely processed into a search plan
	 * except for the (, ), !, and -o operators.  Rearrange the plan so
	 * that the portions of the plan which are affected by the operators
	 * are moved into operator nodes themselves.  For example:
	 *
	 *	[!]--> [-name foo]--> [-print]
	 *
	 * becomes
	 *
	 *	[! [-name foo] ]--> [-print]
	 *
	 * and
	 *
	 *	[(]--> [-depth]--> [-name foo]--> [)]--> [-print]
	 *
	 * becomes
	 *
	 *	[expr [-depth]-->[-name foo] ]--> [-print]
	 *
	 * operators are handled in order of precedence.
	 */

	plan = paren_squish(plan);		/* ()'s */
	plan = not_squish(plan);		/* !'s */
	plan = or_squish(plan);			/* -o's */
	return (plan);
}

/* addPath - helper function used to build a list of paths that were
 * specified on the command line that we are allowed to search.
 */
static char **addPath(char **array, char *newPath)
{
	static int pathCounter = 0;
	
	if (newPath == NULL) {	/* initialize array */
        if ((array = malloc(sizeof(char *))) == NULL) {
			myerr(2, "out of memory");
            return 0;
        }
		array[0] = NULL;
	}
	else {
		array = realloc(array, (++pathCounter + 1) * sizeof(char *));
        if (array == NULL) {
            myerr(2, "out of memory");
            return 0;
        }
		else {
			array[pathCounter - 1] = newPath;
			array[pathCounter] = NULL;	/* ensure array is null terminated */
		}
	}
	return (array);
}

FTS *tree;			/* pointer to top of FTS hierarchy */

/*
 * find_execute --
 *	take a search plan and an array of search paths and executes the plan
 *	over all FTSENT's returned for the given search paths.
 */
int
find_execute(PLAN *plan, char *paths[])
{
	FTSENT *entry;
	PLAN *p;
	int rval;
	char **myPaths;
	int nonSearchableDirFound = 0;
	int pathIndex;
	struct stat statInfo;

	/* special-case directories specified on command line - explicitly examine
	 * mode bits, to ensure failure if the directory cannot be searched
	 * (whether or not it's empty). UNIX conformance... <sigh>
	 */
		
	int strict_symlinks = (ftsoptions & (FTS_COMFOLLOW|FTS_LOGICAL))
	  && COMPAT_MODE("bin/find", "unix2003");

	myPaths = addPath(NULL, NULL);
	for (pathIndex = 0; paths[pathIndex] != NULL; ++pathIndex) {
		int stat_ret = stat(paths[pathIndex], &statInfo);
		int stat_errno = errno;
		if (strict_symlinks && stat_ret < 0) {
		    if (stat_errno == ELOOP) {
                myerrx(1, "Symlink loop resolving %s", paths[pathIndex]);
                return 0;
		    }
		}

		/* retrieve mode bits, and examine "searchable" bit of 
		  directories, exempt root from POSIX conformance */
		if (COMPAT_MODE("bin/find", "unix2003") && getuid() 
		  && stat_ret == 0 
		  && ((statInfo.st_mode & S_IFMT) == S_IFDIR)) {
			if (access(paths[pathIndex], X_OK) == 0) {
				myPaths = addPath(myPaths, paths[pathIndex]);
			} else {
				if (stat_errno != ENAMETOOLONG) {	/* if name is too long, just let existing logic handle it */
					mywarnx("%s: Permission denied", paths[pathIndex]);
					nonSearchableDirFound = 1;
				}
			}
		} else {
			/* not a directory, so add path to array */
			myPaths = addPath(myPaths, paths[pathIndex]);
		}
	}
	if (myPaths[0] == NULL) {	/* were any directories searchable? */
		free(myPaths);
		return(nonSearchableDirFound);	/* no... */
	}

	tree = fts_open(myPaths, ftsoptions, (issort ? find_compare : NULL));
    if (tree == NULL) {
		myerr(1, "ftsopen");
        return 0;
    }
	for (rval = nonSearchableDirFound; (entry = fts_read(tree)) != NULL;) {
		if (maxdepth != -1 && entry->fts_level >= maxdepth) {
            if (fts_set(tree, entry, FTS_SKIP)) {
				myerr(1, "%s", entry->fts_path);
                return 0;
            }
		}

		switch (entry->fts_info) {
		case FTS_D:
			if (isdepth)
				continue;
			break;
		case FTS_DP:
			if (!isdepth)
				continue;
			break;
		case FTS_DNR:
		case FTS_ERR:
		case FTS_NS:
			(void)fflush(stdout);
			mywarnx("%s: %s",
			    entry->fts_path, strerror(entry->fts_errno));
			rval = 1;
			continue;
#ifdef FTS_W
		case FTS_W:
			continue;
#endif /* FTS_W */
		}
#define	BADCH	" \t\n\\'\""
		if (isxargs && strpbrk(entry->fts_path, BADCH)) {
			(void)fflush(stdout);
			mywarnx("%s: illegal path", entry->fts_path);
			rval = 1;
			continue;
		}

		if (mindepth != -1 && entry->fts_level < mindepth)
			continue;

		/*
		 * Call all the functions in the execution plan until one is
		 * false or all have been executed.  This is where we do all
		 * the work specified by the user on the command line.
		 */
		for (p = plan; p && (p->execute)(p, entry); p = p->next);
	}
	free (myPaths);
	finish_execplus();
	if (execplus_error) {
		exit(execplus_error);
	}
    if (errno) {
		myerr(1, "fts_read");
        return 0;
    }
	fts_close(tree);
	return (rval);
}
